File: src/main/java/com/polygloat/model/UserAccount.java
package com.polygloat.model;

import lombok.*;

import javax.persistence.*;
import java.util.HashSet;
import java.util.Set;

@Entity
@Table(uniqueConstraints = {
        @UniqueConstraint(columnNames = {"username"}, name = "useraccount_username"),
        @UniqueConstraint(columnNames = {"third_party_auth_type", "third_party_auth_id"}, name = "useraccount_authtype_auth_id"),

})
@EqualsAndHashCode(onlyExplicitlyIncluded = true, callSuper = true)
@Builder
@AllArgsConstructor
@Data
@ToString(of = {"id", "username", "name"})
public class UserAccount extends AuditModel {

    @Id
    @EqualsAndHashCode.Include
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private String username;

    private String password;

    private String name;

    @OneToMany(mappedBy = "createdBy")
    @Builder.Default
    private Set<Repository> createdRepositories = new HashSet<>();

    @OneToMany(mappedBy = "user")
    private Set<Permission> permissions;

    @Builder.Default
    @Enumerated(EnumType.STRING)
    private Role role = Role.USER;

    @Column(name = "third_party_auth_type")
    private String thirdPartyAuthType;

    @Column(name = "third_party_auth_id")
    private String thirdPartyAuthId;

    @Column(name = "reset_password_code")
    private String resetPasswordCode;


    public UserAccount() {
    }

    public UserAccount(String username) {
        this.username = username;
    }

    public enum Role {
        USER,
        ADMIN
    }
}

File: src/main/kotlin/io/tolgee/model/UserAccount.kt
package io.tolgee.model

import javax.persistence.*

@Entity
@Table(
        uniqueConstraints = [
            UniqueConstraint(
                    columnNames = ["username"],
                    name = "useraccount_username"
            ),
            UniqueConstraint(
                    columnNames = ["third_party_auth_type", "third_party_auth_id"],
                    name = "useraccount_authtype_auth_id")
        ]
)
data class UserAccount(
        @Id
        @GeneratedValue(strategy = GenerationType.IDENTITY)
        var id: Long? = null,
        var username: String? = null,
        var password: String? = null,
        var name: String? = null,
        @Enumerated(EnumType.STRING)
        var role: Role? = Role.USER
) : AuditModel() {

    @OneToMany(mappedBy = "user")
    var permissions: MutableSet<Permission>? = null

    @OneToOne(mappedBy = "userAccount", fetch = FetchType.LAZY, optional = true)
    var emailVerification: EmailVerification? = null

    @Column(name = "third_party_auth_type")
    var thirdPartyAuthType: String? = null

    @Column(name = "third_party_auth_id")
    var thirdPartyAuthId: String? = null

    @Column(name = "reset_password_code")
    var resetPasswordCode: String? = null

    @OneToMany(mappedBy = "user")
    var organizationRoles: MutableList<OrganizationRole> = mutableListOf()

    constructor(id: Long?,
                username: String?,
                password: String?,
                name: String?,
                permissions: MutableSet<Permission>?,
                role: Role?,
                thirdPartyAuthType: String?,
                thirdPartyAuthId: String?,
                resetPasswordCode: String?
    ) : this(id, username, password, name) {
        this.permissions = permissions
        this.role = role
        this.thirdPartyAuthType = thirdPartyAuthType
        this.thirdPartyAuthId = thirdPartyAuthId
        this.resetPasswordCode = resetPasswordCode
    }

    enum class Role {
        USER, ADMIN
    }
}
